### PR TITLE
chore: make workspace template fields nullable

### PR DESCRIPTION
**Pull request by Emyrk**
https://github.com/coder/coder/issues/12359#issuecomment-1971791714

If a user can read a workspace, but not the template, we need to omit those fields returned.

This is an imperfect solution, but atm we just do not return the workspace. 

Some other solutions:

1. Create some "Minimal Template" that a user can read if they can read the workspace. 
2. Just allow read access to the template if you have workspace access.
3. Block workspace access if you do not have template access. (This is a bit complex since rego -> sql cannot take into account 2 resources atm.)

I don't see many good solutions. This PR just nulls out some api fields, and the FE is going to require an update to handle this.

Thoughts?

---

**Note:** If this is the right direction for now:

- FE needs to be updated alongside this
- Unit tests
- Comb it all over

### CODE REVIEW

File: site/src/api/typesGenerated.ts [hunk 1]
```diff
@@ -1640,19 +1640,14 @@ export interface Workspace {
   readonly owner_name: string;
   readonly owner_avatar_url: string;
   readonly organization_id: string;
-  readonly template_id: string;
-  readonly template_name: string;
-  readonly template_display_name: string;
-  readonly template_icon: string;
-  readonly template_allow_user_cancel_workspace_jobs: boolean;
-  readonly template_active_version_id: string;
-  readonly template_require_active_version: boolean;
```
Comment (BrunoQuaresma) [inline site/src/api/typesGenerated.ts:1649]: I always wonder why we just return the template instead of selecting attribute by attribute.

Comment (Emyrk) [reply site/src/api/typesGenerated.ts:1649]: I bet it just grew over time.

**Comment by Emyrk** [comment]
> I don't know, but I'm hesitant about the current solution. I guess it is ok, but it touches a few pages like workspaces list, workspace page, workspace settings pages, and maybe others 🤔. Wondering if having the bug on the pagination count is just better than having a workaround that adds other possible issues. My feeling is, we are moving the problem to another area and not fixing it.

I don't think it's a bad solution, as permission wise, returning a null template is correct. Since you can read the workspace, but not the template.

The problem is that it's currently undefined behavior. At present, you can still interact with the workspace if the endpoints do not interact with the template. So start/stop might work?? I'm not sure.

The status quo of not listing the workspace isn't correct either...

**Comment by BrunoQuaresma** [comment]
Maybe we could have 2 types of template data where one of them can expose some basic fields like template name, display name, and icon so it would not affect the UI too much. So in the UI, we could just gray out the links and actions related to the template. Wdyt?

**Comment by Emyrk** [comment]
> Maybe we could have 2 types of template data where one of them can expose some basic fields like template name, display name, and icon so it would not affect the UI too much. So in the UI, we could just gray out the links and actions related to the template. Wdyt?

I considered that too, I just am unsure if we should even reveal that data if you do not have the perms.

**Comment by sreya** [comment]
@spikecurtis @BrunoQuaresma what do you think should be done here? I'd prefer to not introduce the concept of a `BasicTemplateInfo` or any sort of safe "subset" of template metadata. Do we foresee any sprawling complexity of interpreting `nil` template fields in a workspace response as indicative of a user no longer having access to a template?

**Comment by spikecurtis** [comment]
I can't see any realistic subset of the template that would be useful.  When an admin kicks you off a template, it's _supposed_ to mean you can't use it any more.

I can't comment directly on the front-end complexity, but I think what needs to happen is the workspace is shown in lists of all your workspaces, with the minimal info we can read directly off the workspace including name, creation time, etc.  The only useful action the end user who lost access can do is contact an administrator, so that should be communicated in the front end.

The only reasonable alternative I can see is rearranging the permissions processing so that we can't end up in this state: that is, either destroying the workspace as part of losing template permission, or refusing to remove the permission while a workspace still exists.

**Comment by BrunoQuaresma** [comment]
What we can do in the FE if the template data, like icon and display name, are missing is to display a placeholder:
- 🚫 as icon
- "Template not available" as template name -- Maybe we can think of a better one 🤔

**Pull request closed without merging**